File: src/main/java/intro/to/design/patterns/using/javalang/iterator/BookShelf.java
package intro.to.design.patterns.using.javalang.iterator;

import java.util.ArrayList;

public class BookShelf implements Iterable<Book> {
    private final ArrayList<Book> books;

    public BookShelf(int maxSize) {
        this.books = new ArrayList<>(maxSize);
    }

    public Book getBookAt(int index) {
        return books.get(index);
    }

    public void appendBook(Book book) {
        this.books.add(book);
    }

    public int getLength() {
        return books.size();
    }

    @Override
    public Iterator<Book> iterator() {
        return new BookShelfIterator(this);
    }

    public ArrayList<Book> getBooks() {
        return books;
    }
}

File: c++-design-pattern/src/intro_to_design_patterns_using_java_lang/iterator/BookShelf.cpp
#include "BookShelf.h"

BookShelf::BookShelf() = default;

BookShelf::BookShelf(int maxSize) : BookShelf()
{
    books->reserve(maxSize);
}

BookShelf::~BookShelf() = default;

Book BookShelf::getBookAt(int index) const
{
    return books->at(index);
}

void BookShelf::appendBook(const Book &book)
{
    books->push_back(book);
}

size_t BookShelf::getLength() const
{
    return books->size();
}

const std::shared_ptr<std::vector<Book>> &BookShelf::getBooks() const
{
    return books;
}

std::unique_ptr<BookShelfIterator> BookShelf::iterator()
{
    return std::make_unique<BookShelfIterator>(books);
}
